src/Illuminate/Database/Query/Builder.php

```diff
@@ -339,9 +353,23 @@ protected function parseSub($query)
      */
     public function addSelect($column)
     {
-        $column = is_array($column) ? $column : func_get_args();
+        $columns = is_array($column) ? $column : func_get_args();
+
+        foreach ($columns as $as => $column) {
+            if (is_string($as) && (
+                $column instanceof self ||
+                $column instanceof EloquentBuilder ||
+                $column instanceof Closure
+            )) {
+                if (is_null($this->columns)) {
+                    $this->select($this->from.'.*');
+                }
 
-        $this->columns = array_merge((array) $this->columns, $column);
+                $this->selectSub($column, $as);
+            } else {
+                $this->columns = array_merge((array) $this->columns, [$column]);
```

Comment:
```suggestion
                $this->columns[] = $column;
```

Reply:
Umm yup, so obvious. I did that in `select()`. Just missed it here. Thanks!